cilium-cli/connectivity/tests/ipsec_xfrm.go

```diff
@@ -129,3 +130,263 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+	}
+
+	// Extract authentication key
+	if match := regexp.MustCompile(`auth-trunc \S+ 0x([0-9a-fA-F]+)`).FindStringSubmatch(block); len(match) >= 2 {
+		state.AuthKey = match[1]
+	} else if match := regexp.MustCompile(`auth \S+ 0x([0-9a-fA-F]+)`).FindStringSubmatch(block); len(match) >= 2 {
+		state.AuthKey = match[1]
+	}
+
+	// Extract encryption key
+	if match := regexp.MustCompile(`enc \S+ 0x([0-9a-fA-F]+)`).FindStringSubmatch(block); len(match) >= 2 {
+		state.CryptKey = match[1]
+	}
+
+	// Extract AEAD key (for GCM modes)
+	if match := regexp.MustCompile(`aead \S+ 0x([0-9a-fA-F]+)`).FindStringSubmatch(block); len(match) >= 2 {
+		state.AeadKey = match[1]
+	}
+
+	return state, nil
+}
+
+// isCiliumManagedState checks if an XFRM state is managed by Cilium
+func (t *ipsecKeyDerivationTest) isCiliumManagedState(block string) bool {
+	// Cilium uses reqid 1 for its XFRM states
+	return strings.Contains(block, "reqid 1")
+}
+
+// validateKeyDerivation validates that keys are correctly derived and consistent across nodes
+func (t *ipsecKeyDerivationTest) validateKeyDerivation(testContext *check.Test, nodeStates map[string][]xfrmStateInfo) {
+	nodeIPs := make([]string, 0, len(nodeStates))
+	for nodeIP := range nodeStates {
+		nodeIPs = append(nodeIPs, nodeIP)
+	}
+
+	if len(nodeIPs) < 2 {
+		testContext.Logf("Not enough nodes with XFRM states for validation (have %d, need 2)", len(nodeIPs))
+		return
+	}
+
+	validationCount := 0
+
+	// Compare states between every pair of nodes
+	for i := 0; i < len(nodeIPs); i++ {
+		for j := i + 1; j < len(nodeIPs); j++ {
+			node1IP := nodeIPs[i]
+			node2IP := nodeIPs[j]
+
+			node1States := nodeStates[node1IP]
+			node2States := nodeStates[node2IP]
+
+			testContext.Logf("Validating key derivation between nodes %s and %s", node1IP, node2IP)
+
+			// Find matching tunnels between the two nodes
+			matches := t.findMatchingTunnels(node1States, node2States, node1IP, node2IP)
+			if len(matches) == 0 {
+				testContext.Logf("No matching tunnels found between nodes %s and %s", node1IP, node2IP)
+				continue
+			}
+
+			// Validate each matching tunnel pair
+			for _, match := range matches {
+				t.validateTunnelKeyPair(testContext, match.state1, match.state2, node1IP, node2IP)
+				validationCount++
+			}
+		}
+	}
+
+	if validationCount == 0 {
+		testContext.Fatalf("No tunnel key pairs found for validation - this indicates a problem with the test or IPsec setup")
+	}
+
+	testContext.Logf("Successfully validated %d tunnel key pairs", validationCount)
+}
+
+// tunnelMatch represents a matched tunnel pair between two nodes
+type tunnelMatch struct {
+	state1 xfrmStateInfo
+	state2 xfrmStateInfo
+}
+
+// findMatchingTunnels finds XFRM states that represent the same tunnel between two nodes
+func (t *ipsecKeyDerivationTest) findMatchingTunnels(states1, states2 []xfrmStateInfo, node1IP, node2IP string) []tunnelMatch {
+	var matches []tunnelMatch
+
+	for _, state1 := range states1 {
+		for _, state2 := range states2 {
```

Comment:
Could we index the states for each node by their `Src` and `Dst`? That way we'd avoid this second `for` loop.